### PR TITLE
[WIP] Support Multi-Blank Transducer in Espnet2

### PR DESCRIPTION
**Pull request by jctian98**
This PR attempts to support multi-blank transducer in Espnet 2

Paper related: https://arxiv.org/abs/2211.03541

Acknowledgment: The original implementation is from NeMo (https://github.com/NVIDIA/NeMo) and by @hainan-xv. The original PR in NeMo is https://github.com/NVIDIA/NeMo/pull/5527

### CODE REVIEW

**Comment by jctian98** [comment]
Still in progress. I'll report some results on Librispeech when this PR is fully validated.

**Comment by jctian98** [comment]
The current version is validated on Librispeech dataset. Below are the results. 
Experiments suggest the Multi-Blank Transducer criterion outperforms the vanilla transducer in speed and accuracy, which is well-aligned with the paper. 

| Criterion | Big blanks | Multi_blank_sigma | Test-clean WER | Test-other WER | RTF |
|-----------|-------------|---------------------|------------------|-------------------|-----|
|Vanilla Transducer | - | - | 2.9 | 6.7 | 0.461 |
| Multi-Blank Transducer | [2,4,8] | 0.05 | 2.8 | 6.6 | 0.324 |
| Multi-Blank Transducer | [2,3,4,5,6,7,8] | 0.05 | 2.8 | 6.4 | 0.325 | 

Several hints for this method:
1. Currently the multi-blank transducer method is only compatible with greedy search.
2. The multi-blank transducer requires more training epochs. In the standard Espent recipe, we train for 25 epochs. But in these experiments, the number of epochs is 50.
3. In the internal implementation of the multi-blank transducer, the order of big_blanks would be reversed (which is not expected). E.g., if the multi_blank transducer accepts the argument `big_blank_durations`  `[2,4,8]`, it actually considers the sequence as `[8, 4, 2]`. Although this has been handled in this PR, making a record is good.

I am happy to collaborate so that this PR can pass the CI.

**Review by b-flo — CHANGES_REQUESTED** [review]
Hi @jctian98 ,

It would be better to move the `rnnt_multi_blank` folder under `asr` instead of `asr_transducer`. 

The second folder is reserved for the standalone version of ESPnet2-Transducer while the first is for the shared version with CTC/Att, which you're working with.

File: espnet2/asr/espnet_model.py [hunk 1]
```diff
@@ -110,15 +112,25 @@ def __init__(
...
 
-            self.criterion_transducer = RNNTLoss(
-                blank=self.blank_id,
-                fastemit_lambda=0.0,
-            )
+            if len(transducer_multi_blank_durations) == 0:
+                from warprnnt_pytorch import RNNTLoss
+                self.criterion_transducer = RNNTLoss(
+                    blank=self.blank_id,
+                    # fastemit_lambda=0.0,
```
Comment (b-flo) [inline espnet2/asr/espnet_model.py:122]:
```suggestion
                    fastemit_lambda=0.0,
```

**Comment by hainan-xv** [comment]
> The current version is validated on Librispeech dataset. Below are the results. Experiments suggest the Multi-Blank Transducer criterion outperforms the vanilla transducer in speed and accuracy, which is well-aligned with the paper.
> 
> Criterion	Big blanks	Multi_blank_sigma	Test-clean WER	Test-other WER	RTF
> Vanilla Transducer	-	-	2.9	6.7	0.461
> Multi-Blank Transducer	[2,4,8]	0.05	2.8	6.6	0.324
> Multi-Blank Transducer	[2,3,4,5,6,7,8]	0.05	2.8	6.4	0.325
> Several hints for this method:
> 
> 1. Currently the multi-blank transducer method is only compatible with greedy search.
> 2. The multi-blank transducer requires more training epochs. In the standard Espent recipe, we train for 25 epochs. But in these experiments, the number of epochs is 50.
> 3. In the internal implementation of the multi-blank transducer, the order of big_blanks would be reversed (which is not expected). E.g., if the multi_blank transducer accepts the argument `big_blank_durations`  `[2,4,8]`, it actually considers the sequence as `[8, 4, 2]`. Although this has been handled in this PR, making a record is good.
> 
> I am happy to collaborate so that this PR can pass the CI. @ftshijt

Congrats on making this work with ESPnet and thanks!

**Comment by sw005320** [comment]
> Hi @jctian98 ,
> 
> It would be better to move the `rnnt_multi_blank` folder under `asr/transducer` instead of `asr_transducer`.
> 
> The second folder is reserved for the standalone version of ESPnet2-Transducer while the first is for the shared version with CTC/Att, which you're working with.

I agree

**Comment by sw005320** [comment]
Can you apply black?
https://github.com/espnet/espnet/actions/runs/4132028864/jobs/7140252200#step:8:22

**Comment by jctian98** [comment]
@b-flo 
> P.S: Unrelated but is the RTF you reported for vanilla Transducer with greedy search also? It seems really slow.

All decoding is with greedy search. It is slow because I use `$infernece_nj=88`. After setting `$infernece_nj=20`, the three RTF numbers become `0.190`, `0.152` and `0.157` respectively.  (I am not very sure how many physical CPU cores are on my machine)

The implementation code of the multi-blank transducer is moved from `asr_transducer` to `asr/transducer`

@sw005320 
> Can you apply black?

Didn't know it. Have applied now.

@hainan-xv 

> Congrats on making this work with ESPnet and thanks!

Thanks for your kind support!

**Comment by b-flo** [comment]
I'll review the PR in a few days, thanks again for this contribution @jctian98 

> All decoding is with greedy search. It is slow because I use $infernece_nj=88. After setting $infernece_nj=20, the three RTF numbers become 0.190, 0.152 and 0.157 respectively. (I am not very sure how many physical CPU cores are on my machine)

OK great, thank you! You can look at `/proc/cpuinfo` or use `top` (press `1` to show CPU cores). I assume the `inference_nj` value is too high given the RTF difference.

Btw, do you have WER and RTF for the vanilla Transducer with default beam search, please?

**Comment by jctian98** [comment]
Here I update the experimental results like below:
| Criterion | Big blanks | Multi_blank_sigma | Test-clean WER | Test-other WER | RTF (nj=20) |
|-----------|-------------|---------------------|------------------|-------------------|-----|
|Vanilla Transducer - Greedy Search | - | - | 2.9 | 6.7 | 0.190 |
|Vanilla Transducer - Default Beam Search (beam=10) | - | - | 2.8 | 6.5 | 0.750 |
| Multi-Blank Transducer - Greedy Search | [2,4,8] | 0.05 | 2.8 | 6.6 | 0.152 |
| Multi-Blank Transducer - Greedy Search | [2,3,4,5,6,7,8] | 0.05 | 2.8 | 6.4 | 0.157 | 

I have checked `/proc/cpuinfo` and it suggests I have 88 CPU cores. Maybe some competition of system resources would slow down the RTF when `$inference_nj=88`. After changing from `$inference_nj=20` to `$inference_nj=10`, the RTF gets better only slightly. So I suppose `$inference_nj=20` is valid on my machine. 

I have also report the WER and RTF for vanilla transducer beam search. No LM used.

I have removed `asr_transducer/multi_blank_rnnt` folder. Sorry for this mistake.

CI suggests the `black` should be applied to two files. But it seems applying `black` to these files on my local machine doesn't change them. Sorry I don't know how to get through this. Please help. Thanks! :)

File: espnet2/asr/transducer/rnnt_multi_blank/utils/rnnt_helper.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    if idx < length:
+        copy_data_1d(source, dest, idx)
+        dest[idx] *= -1.0
+        dest[idx] *= 1.0 + fastemit_lambda
+
+
+def get_workspace_size(
+    maxT: int, maxU: int, minibatch: int, gpu: bool
+) -> Tuple[Optional[int], global_constants.RNNTStatus]:
+
```
Comment (b-flo) [inline espnet2/asr/transducer/rnnt_multi_blank/utils/rnnt_helper.py:123]:
```suggestion
```

File: espnet2/bin/asr_inference.py [hunk 3]
```diff
@@ -175,12 +176,24 @@ def __init__(
 
         # 4. Build BeamSearch object
         if asr_model.use_transducer_decoder:
+
```
Comment (b-flo) [inline espnet2/bin/asr_inference.py:179]:
```suggestion
```

**Review by b-flo — COMMENTED** [review]
> CI suggests the black should be applied to two files. But it seems applying black to these files on my local machine doesn't change them. Sorry I don't know how to get through this. Please help. Thanks! :)

Below are the CI flagged issues in your PR. There are also 2 more in `asr_inference.py` `l.419`  and `l.686`:

```python3
        else:                                                                                        
                                                                                                     
            # Normal ASR
```

You can skip them in this PR I think.

**Comment by b-flo** [comment]
> I have checked /proc/cpuinfo and it suggests I have 88 CPU cores. Maybe some competition of system resources would slow down the RTF when $inference_nj=88. After changing from $inference_nj=20 to $inference_nj=10, the RTF gets better only slightly. So I suppose $inference_nj=20 is valid on my machine.

Maybe some bottleneck but I'll take a look. Did you use `calculate_rtf` utils?

> I have also report the WER and RTF for vanilla transducer beam search. No LM used.

Thank you for the report! Well, that's a rather high RTF!
It may be worth further investigation. I'm currently using a somewhat similar architecture and same decoding parameters for `Libri-100` and end-up with an RTF of ~0.246 (~0.1 for `ALSD` and `mAES` for same perfs). Without taking into account difference between the shared/standalone versions, It may help us improve some parts.

**Comment by jctian98** [comment]
@b-flo 
I am reporting the results from `calculate_rtf` utils.

In `path.sh`  I have `export OMP_NUM_THREADS=1`; GPU model is `Intel(R) Xeon(R) Platinum 8255C CPU @ 2.50GHz`. The `cpuinfo` file is attached for inspection. [cpuinfo.txt](https://github.com/espnet/espnet/files/10709615/cpuinfo.txt) 

I am happy to help solve the speed bottleneck. Please let me know if you need any more info. Btw, I am working in a docker container created by a large GPU cluster so the physical hardware is always changing (but should have similar hardware configurations).

I have merged the current espnet but there are still some problems about formatting in CI. I'll be off for 2 days and will fix these problems on Monday. :)

**Review by b-flo — CHANGES_REQUESTED** [review]
Good work @jctian98. I didn't review all the code in `rnnt_multi_blank` but the rest is OK outside some minor things.

That's being said, but I'm wondering if `rnnt_multi_blank` should be an external dependency alongside `warp-transducer`. Or at least, it should be put somewhere that make more "sense".
It's for another PR but I think we should discuss what to do with such implementations. @sw005320 Do you have an opinion on that?

(Or we could also extend `warp-transducer` for multi-blank and remove numba dependency with that.)

File: espnet2/asr/transducer/beam_search_transducer.py [hunk 4]
```diff
@@ -883,3 +902,55 @@ def modified_adaptive_expansion_search(
...
+
+        for t, enc_out_t in enumerate(enc_out):
+            # case 1: skip frames until big_blank_duration == 1
+            if big_blank_duration > 1:
+                big_blank_duration -= 1
+                continue
+
+            symbols_added = 0
+            while symbols_added <= 3:
+                dec_out, state, _ = self.decoder.score(hyp, cache)
```
[the remark this answers is not in the document]
Comment (jctian98) [reply espnet2/asr/transducer/beam_search_transducer.py:940]: This is a greedy search so I don't think LM would help.

[the remark this answers is not in the document]
Comment (b-flo) [reply espnet2/asr/transducer/beam_search_transducer.py:940]: I don't mind but you should notify the user, during beam init, if `mgb` is used and `lm is not None` then!

[the remark this answers is not in the document]
Comment (jctian98) [reply espnet2/asr/transducer/beam_search_transducer.py:936]: I see.
I suppose a reminder is also not provided for the plain greedy search. So I added a warning when `beam_size=1` and `use_lm=True`.

**Review by b-flo — APPROVED** [review]
LGTM. It can be merged after CI fixes (most of them come from loss computation files).

**Comment by sw005320** [comment]
> Good work @jctian98. I didn't review all the code in `rnnt_multi_blank` but the rest is OK outside some minor things.
> 
> That's being said, I'm wondering if `rnnt_multi_blank` should be an external dependency alongside `warp-transducer`. Or at least, it should be put somewhere that make more "sense". It's for another PR but I think we should discuss what to do with such implementations. @sw005320 Do you have an opinion on that?

I agree.
Yes, we can think of it in another PR.
Transducer is also used in speech translation and it is better to put it in a better place and we can put some custom losses (or make it as an external tool).
I'm thinking of putting this in `espnet/espnet/tree/master/espnet2/loss/rnnt_multi_blank` or something like that.

> (Or we could also extend `warp-transducer` for multi-blank and remove numba dependency with that.)

This is a nice idea.
We can add this to our TODO list.

**Comment by sw005320** [comment]
@jctian98, I think the current CI issue comes from `flack8`
Can you check it?

**Comment by jctian98** [comment]
Hi, sorry for my late reply.

(1) On my local machine, the command `flake8 espnet2/asr/transducer/rnnt_multi_blank/ espnet2/asr/espnet_model.py espnet2/tasks/asr.py espnet2/bin/asr_inference.py` gives no output so I suppose I have passed the flake8 locally with current commit.

(2) In addition, when running `ci/test_flake8.sh`, the output is
```
bash ci/test_flake8.sh 
flake8-docstrings ready files coverage: 178 / 234 = 76.0683%
espnet/nets/beam_search_timesync.py:195:44: E741 ambiguous variable name 'l'
        hyps = sorted(new_hyps, key=lambda l: scores[l], reverse=True)[: self.beam_size]
```
But the file mentioned above is not modified in this PR.

(3) In the CI check log, there are some comments suggesting `multi line docstring summary not separated with an empty line`. But the empty lines exactly exist after the multi-line doc strings. (e.g., `espnet2/asr/transducer/rnnt_multi_blank/utils/cpu_utils/cpu_rnnt.py:338`)

I'm a bit confused about the current status. Please let me know what should I do next.

**Comment by sw005320** [comment]
Can you check the versions of the test tools (used in your local and used in C)?

**Comment by sw005320** [comment]
> Sorry, I don't understand what the versions used in C are.

used in CI

**Comment by sw005320** [comment]
image

**Comment by jctian98** [comment]
Hi，

with `flake8=5.0.4`, `ci/test_flake8.sh` also gives no error message on my machine.

following [this](https://github.com/espnet/espnet/actions/runs/4194631493/jobs/7273051194#step:6:11875), I have also changed to `flake8=3.9.2`. `ci/test_flake8.sh` also gives no error message :(

**Comment by sw005320** [comment]
@ftshijt, do you know what’s happening?

**Comment by sw005320** [comment]
@jctian98, we need to add
- unit test https://github.com/espnet/espnet/tree/master/test/espnet2/asr
- integration test https://github.com/espnet/espnet/blob/master/ci/test_integration_espnet2.sh
- upload a model and corresponding results to egs2/librispeech/asr1/README.md

**Comment by jctian98** [comment]
> @jctian98, we need to add
> 
> * unit test https://github.com/espnet/espnet/tree/master/test/espnet2/asr
> * integration test https://github.com/espnet/espnet/blob/master/ci/test_integration_espnet2.sh
> * upload a model and corresponding results to egs2/librispeech/asr1/README.md

Will work on it soon!

**Comment by jctian98** [comment]
This is a follow-up.

1. The previous test code does not address the transducer model defined in `espnet2/asr/espnet_model.py` (although it tests the standalone transducer `espnet2/asr_transducer`). This PR adds new code to test both the vanilla and Multi-Blank transducer.
2. the Multi-Blank Transducer criterion only runs on GPU. It cannot be tested if the test environment has no GPU (like now, maybe).
3. `codecov/patch` reports many cautions for the GPU-version Multi-Blank transducer implementation. This part of code cannot be covered unless a GPU is accessible. This is the only test that fails in CI.

Unit Test: see `test/espnet2/asr/test_espnet_model.py` and `test/espnet2/asr/transducer/test_transducer_beam_search.py`
Integration test: see `ci/test_integration_espnet2.sh`
Again, for both unit and integration tests, some code will not run unless a GPU is accessible. See some `if` in the code.

I suppose the test code may need an additional review.

The model has been uploaded to HF.

**Comment by sw005320** [comment]
Thanks!
They look good to me, considering the current mult-blank transducer loss is only available for CUDA.

@hainan-xv, in the future, it would be great if you also prepare the CPU version for debugging and CI purposes.

**Pull request merged**